core/drivers/crypto/versal/hash.c

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2022
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <assert.h>
+#include <drvcrypt.h>
+#include <drvcrypt_hash.h>
+#include <initcall.h>
+#include <ipi.h>
+#include <kernel/panic.h>
+#include <kernel/refcount.h>
+#include <kernel/spinlock.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+
+/* single engine */
+struct mutex lock = MUTEX_INITIALIZER;
+
+#define FIRST_PACKET	BIT(30)
+#define NEXT_PACKET	BIT(31)
+
+enum versal_sha3_state {
+	SHA3_STNDBY = 0, SHA3_INIT, SHA3_RUN
+};
+
+struct update_request {
+	uint8_t *data;
+	size_t len;
+	STAILQ_ENTRY(update_request) link;
+};
+
+struct versal_hash_ctx {
+	struct crypto_hash_ctx hash_ctx;
+	enum versal_sha3_state state;
+	STAILQ_HEAD(update_request_list, update_request) req_list;
```

Comment:
The `STAILQ_HEAD()` macro expands to:
``` C
struct update_request_list { ... }
```
I would prefer to have this done at the top level instead of inside `struct versal_hash_ctx` in case you ever need to use `struct update_request_list` at another place.